sirepo/auth/ldap.py

```diff
@@ -0,0 +1,96 @@
+"""LDAP authentication support
+:copyright: Copyright (c) 2022-2023 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern import pkconfig
+from pykern import pkinspect
+from pykern.pkdebug import pkdlog, pkdp
+from pykern.pkcollections import PKDict
+import ldap
+import re
+import sirepo.quest
+import sirepo.util
+
+AUTH_METHOD = "ldap"
+
+AUTH_METHOD_VISIBLE = True
+
+_cfg = None
+
+# ldap dn requires escaped special chars
+_ESCAPE_DN_MAIL = re.compile(r"([,\\#<>;\"=+])")
+
+# windows login dialogue char limit for passwords
+_MAX_ENTRY = 127
+
+#: module handle
+this_module = pkinspect.this_module()
+
+#: Well known alias for auth
+user_model = "AuthEmailUser"
+
+
+class API(sirepo.quest.API):
+    @sirepo.quest.Spec("require_cookie_sentinel")
+    def api_authLdapLogin(self):
+        def _bind(creds):
+            try:
+                ldap.initialize(_cfg.server).simple_bind_s(creds.dn, creds.password)
+            except Exception as e:
+                m = "Unable to contact LDAP server"
+                if isinstance(e, ldap.INVALID_CREDENTIALS):
+                    m = "Incorrect email or password"
+                else:
+                    pkdlog("email={} exception={}", creds.email, e)
+                raise sirepo.util.Response(self.reply_ok(PKDict(formError=m)))
+
+        def _user(email):
+            m = self.auth_db.model(user_model)
+            u = m.unchecked_search_by(unverified_email=email)
+            if not u:
+                u = m.new(unverified_email=email, user_name=email)
+                u.save()
+            return u
+
+        def _validate_and_escape_creds(req):
+            email = req.req_data.email.lower()
+            password = req.req_data.password
+            if len(email) > _MAX_ENTRY or len(password) > _MAX_ENTRY:
+                raise sirepo.util.UserAlert(
+                    "invalid user and/or password",
+                    f"email={email} or password greater than 127 chars",
+                )
+            if len(_cfg.dn_suffix) > _MAX_ENTRY:
```

Comment:
This test is an invariant. It's part of cfg. Instead of "str" for pkconfig.init, create a function that tests the value. For examples search around for "def _cfg_" which are examples of pkconfig validators.

From a security perspective, do not return configuration errors to the user. Security permeates code. It's things like this that slip by reviewers (as it did the first time with me). You never want to return a configuration knowledge (invalid dn is knowledge) to an end user unless it is completely safe to do so.

Reply:
Unsure if done correctly, replaced 'str' in 'dn_suffix' with function that raises AssertionError similar to pkconfig.raise_error().